R/FilteredData.R

```diff
@@ -1223,6 +1262,162 @@ FilteredData <- R6::R6Class( # nolint
           }
         })
       })
+    },
+
+    # Snapshot manager module.
+    # @description
+    # Controls capturing, restoring, and saving filter state snapshots.
+    ui_snapshot_manager = function(id) {
+      ns <- NS(id)
+      div(
+        class = "snapshot_manager_content",
+        div(
+          class = "snapshot_table_row",
+          span(tags$b("Snapshot manager")),
+          actionLink(ns("snapshot_add"), label = NULL, icon = icon("camera"), title = "add snapshot"),
+          actionLink(ns("snapshot_reset"), label = NULL, icon = icon("undo"), title = "reset initial state"),
+          NULL
+        ),
+        uiOutput(ns("snapshot_list"))
+      )
+    },
+    # Snapshot manager module.
+    # @description
+    # Captures, stores, restores, and saves filter state snapshots.
+    # Creates UI elements for each captured snapshot.
+    # @details
+    # Snapshot is a `teal_slices` converted to a list of list of lists.
+    # Attributes other than class remain unchanged.
+    # Snapshots are added on user request and stored in a `reactiveVal`.
+    # Initial state application is the first snapshot.
+    # Snapshots are displayed in a table that shows the name, an select button and a save button.
+    # Selecting rebuilds `teal_slices` from snapshot, clears filter states and sets new state.
+    # Saving rebuilds `teal_slices` from snapshot and saves it to json file.
+    # Initial state is not explicitly enumerated in table, has separate reset button.
+    srv_snapshot_manager = function(id) {
+      moduleServer(id, function(input, output, session) {
+        ns <- session$ns
+
+        # Store global filter states.
+        snapshot_history <- reactiveVal({
+          list(
+            "Initial application state" = disassemble_slices(isolate(self$get_filter_state()))
+          )
+        })
+
+        # Snapshot current application state - name snaphsot.
+        observeEvent(input$snapshot_add, {
+          showModal(
+            modalDialog(
+              textInput(
+                session$ns("snapshot_name"),
+                "Name the snapshot",
+                width = "100%",
+                placeholder = "Meaningful, unique name"
+              ),
+              footer = tagList(
+                actionButton(session$ns("snapshot_name_accept"), "Accept", icon = icon("thumbs-up")),
```

Comment:
Please use `ns` everywhere in this module - as it is in teal::srv_snapshot_manager
```suggestion
                actionButton(ns("snapshot_name_accept"), "Accept", icon = icon("thumbs-up")),
```

Reply:
Gladly! I thought there was a convention module servers within `R6` classes usually (always?) use `session$ns`.

Reply:
No convention, I just checked diffs between this module and one in teal and it was one of them